### ISSUE TITLE
Redefine strongly typed containers as a more flexible, repeatable, embeddable construct.

### ISSUE THREAD

**Issue by ghost** [thebuzzmedia/universal-binary-json#48]
# What we have...

Draft 11 added support for Strongly Typed Containers (Array and Object) - http://ubjson.org/type-reference/container-types/#optimized-format

The underlying premise of the design was to define a header (TYPE and/or COUNT) at the beginning of a container which contains X number of Value Type elements.
# Short-comings...

The biggest problem with this approach is what has been getting discussed in #43 - strongly typed containers cannot themselves contain other containers (strongly typed or not) -- so this construct, while highly optimized for certain use-cases, introduces a limitation that doesn't exist in JSON (any container can contain any type -- even if they are not of the same type).
# Ok, so what about Issue 43?

Issue #43 fixes that limitation, but as been pointed out to me privately by @kxepal - we still find ourselves with some limitations of STCs the way they are defined -- they are still too rigid and not for a great technical reason.
# What limitations still exist?

Namely:
1. Mixed-type containers cannot be optimized - you either need to use the largest type to store all the values (e.g. in the case of numbers) or you have to rely on a standard un-type-optimized container.
2. The COUNT provided in an STC right now dictates the hard end of that container, e.g. if an array has a count of `4` that means after the 4th element the scope of that array closes. This can potentially rule out the use of highly optimized STCs in streaming environments (like a DB or large client-server requests) where the response cannot be totaled ahead of time -- the workaround for this currently is to change the structure of your data to consume 0 or more containers and append them together into a single datum - this is frustrating for implementors (needing to change data to meet limitations of spec) or a non-starter for using STCs in these cases.
3. JSON natively supports mix-typed containers (e.g. arrays containing strings and objects and other arrays and numbers) - with Draft 11 versions of STC, UBJSON imposed a new limitation specific to UBJSON that typed containers could only contain a single type - requiring implementors that had to deal with mix-typed containers to either remodel their data or add app logic to work around the limitations in the spec (same problem as the previous point)
# How do we address these?

This proposal is taking the finalized header defined in #43 (here - https://github.com/thebuzzmedia/universal-binary-json/issues/43#issuecomment-48664267) that looks like this:

```
[{]                                 // this is an object
    [#][i][3][$][[]                 // of arrays (3 of them)
        [#][i][3][$][{]             // of objects (3 of them)
            [#][i][3][$][[]         // of arrays (3 of them... again)
                [#][i][3][$][i]     // of int8s (3 of them)
```

and combining it with an idea proposed by @kxepal -- allowing the header to be repeated 0 or more times within a container, for example:

```
[[]                                 // array start...
    [#][i][3][$][i]                 // 3x int8s to follow...
    [1]
    [2]
    [3]
    [#][i][3][$][d]                 // 2x float32 to follow...
    [1.11]
    [2.22]
    [#][i][2]                       // 2 fully typed values to follow...
    [S][i][3][bob]
    [L][9223372036854775807]
[]]
```
# Changes required to support this...

Roughly:
- `#` or `#$` (or the combination of the two) would constitute a "header" section inside of a container. Encountering those markers inside a container convey information about the data that is going to follow until the container ends or the next 'header' marker is hit. This is no longer a container header but rather a header for a run of data that follows it.
- The `]` or `}` marker characters would ALWAYS have to be used now to indicate the end of the scope of a container, because the count in the header of a container no longer indicates when parsing of that scope ends.
# Implications to use

This would support everything from using STCs the way they are defined now to the more extreme case of optimized/mixed-type containers that we never could support before (but are supported in existing in JSON)
# Down sides

Currently the only downside I can think of here is the actual implementation of parsing a strongly typed, mixed-typed container in a language like Java or C# (C++?) where there is no longer a guarantee that parsing  a list of int32 values can be safely represented as a `List<Integer>` but must instead be some UBJSON super type or worse a `List<Object>` which is a no-go.

I don't think dynamically typed languages will suffer from this, but in strongly typed languages it is sort of nasty.

As the spec goes, I think this is a nice addition to the flexibility and the _academically_ correct way to define strongly typed containers -- in reality though, maybe supporting them in generation/parsing is so painfully inefficient that it doesn't make sense.

Thoughts?

**Comment by MikeFair** [comment]
In reading through everything there's a lot of competing demands.  The desires are to keep it small and fast (aka optimized) and at the same time deal with JSONs arbitrariness without having to degrade everything in strongly typed languages.  

Something I think that's called for here to keep everything optimized is a special JSONArray class for accessing data through these strongly typed containers (or "Segments" as I've been calling them).  

The only caveat is creating base classes for JSONValue and JSONSegment for each of the more specific types (JSONSegmentUInt8, JSONSegmentFloat64, JSONSegmentNull, etc) or the non-object oriented equivalents for those languages.  

---

An example of the decoder process would go something similar to this:
The decoder found 200 [Z] starting at index 500 in an array (aka somewhere in the middle):

I've been suggesting the UBJSON for 200 nulls as the following:
[[] .... #[200][Z] .... []]
(Use the length descriptor from #64; dropped the [$] as it added no value, [#] is restricted to only specify the count for a single [type].  This is what I'm referring to as a Segment.)

The psuedocode for creating 200 null values starting at index 500 would be:
 newSegment = new JSONNullSegment(200, 500);

This creates a new instance of the JSONNull variant of the JSONSegment class, and specifies its length as 200 and starting index at 500.  (The JSONNull class happens to be extremely simple, because it has no need to store any data to speak of, it can only ever return the same "null" value).

Now, this segment can be merged into the JSONArray container by calling something like:
jsonArray.Add(newSegment).  

Assuming the JSONArray is a List of JSONSegment, Add loops through each Segment in its list finding where element 500 should be.  It will then do what it needs to do in order to insert the Segment and make sure the startIndex in the neighboring Segments say the right things (at worst it would need to split an existing Segment into two so it can insert the new values between them; also if the jsonArray currently isn't long enough it can create a JSONNullSegment to fill in the gaps).

---

Now when something wants to get at element 650 (which is part of this null segment) it will call jsonArray[650]; this causes the JSON Array to iterate through its segments looking for the segment that contains index 650 and asks that segment to return the JSONValue stored at position 650 (which it can do because it knows that it starts at 500).

That eliminates the problems with the compiler and strong type segments and prevents reducing it to using a List of Objects.  The JSONTypeSegments know how to store and retrieve JSONValues from the efficient structures that deal exclusively in their specialization.

**Comment by MikeFair** [comment]
I forgot to mention in my last post, about how this also helps with the process of generating the UBJSON for these segments to begin with.  

Assume we took an original 10 element JSON Array like this and read it in:
[0, 1, 2, 3, null, 5, "six", "seven", "eight", 9.0, 10.0, 11.0]

Got "0"; jsonArray.Add(new JSONUint8(0)) //Using 1's based indexing
Got "1"; jsonArray.Add(new JSONUint8(1))
Got "2"; jsonArray.Add(new JSONUint8(2))
Got "3"; jsonArray.Add(new JSONUint8(3))
Got "null"; jsonArray.Add(new JSONNull())
... and so like this adding every element as a single JSONValue to the jsonArray ...

Two things can happen here, when adding a JSONValue, a really smart Add function could just merge the elements into an existing Segment as they are being added.  

A dumb Add function could just add new Values for every element (the equivalent of a List of Object).
However, once the dumb function is done adding the elements, the parser could call jsonArray.reduce().  The reduce function would then merge elements together into their appropriate Segments and combine neighboring Segments if they were of the same type.

In this way, creating the optimized list of segments almost takes care of itself because we have this smart jsonArray class.

Then every segment has a toUBJ() function similar to the common toString().

Since the Segments are strongly typed, they can each know how to produce their own Segment code.
For Segments of length 0, 1, or 2 it will return Empty, [type][value], or [type][value][type][value] because there's no optimization for under 3 consecutive values.  For longer Segments, they can easily produce the [#][LENGTH][VALUES] output.

**Comment by Miosss** [comment]
@MikeFair 
I think we should not enforce any implementation decisions, especially such general ones.

For me, optimized format of containers (whether it is STC, ND, typespec, or whatever) is strictly feature of UBJSON and should NOT be exposed in APIs and especially it should not be default.

The basic reason is JSON - it does not have anything like "optimized" containers. It has plain containers. If I silently use UBJ as a JSON-transport protocol (optimized form over the wire) then neither side of communication is interested in internals of UBJ - they want to write JSON and we only use more optimized protocol for transmission. 

The second reason is simplicity/efficiency. Yes, array of 500 nulls can be effectively optimized out as almost 0 memory structure. But usually parties transmit more than just nulls (some actual data maybe?). If they transmit two "segments", firstly 1024 integers and then 1024 strings, it does not change anything in parsed structure, whether you use array of two segments (like you propose) or array of some mixed-type values. This is in sense of memory footprint. But efficiency drops in the first example, because you have structure of segments and elements, and in the second case, you have only elements - browsing will be faster, even in random-access (if implemented properly).

When I parse UBJ. I am absolutely not interested in what it looked like before - I want the data. I trust UBJ, and library that I use, that constructed UBJ was optimized as much as possible, of course concerning time of the encoding/decoding as well.

The only thing I care about when enc/decoding is what logical structure my message will have - if I write app, I must specify how my communication will look like. API which I use must give me possibilities to write logically messages that I want, and this structure is identical in UBJ and in JSON. How will it be encoded is not really important, if it is done properly, effectively and can be reversed.

**Comment by MikeFair** [comment]
@Miosss
As I read through the comments some of them are exactly about implementation details like this.  I think it's a fair concern to talk about actually have to code it.  By declaring an expectation that a UBJ implementation is expected to handle arrays of arbitrary data with strongly typed segments (at least for decoding them, but also for creating them) we declare consistancy with JSON arrays which are like that.  Some of the proposals/comments do not work because they attempt to create a JSON Array that can only be optimized if the whole array is a single type.  Or are assuming the end user is going to request something that isn't a JSON typed thing.

> I think we should not enforce any implementation decisions, especially such general ones.

> For me, optimized format of containers (whether it is STC, ND, typespec, or whatever) is strictly feature of UBJSON and should NOT be exposed in APIs and especially it should not be default.

The comment wasn't looking to enforce anything but instead create a context for actually creating optimized containers in practice.  How would you do it?  

It's one thing to say "if the parser detects xyz" it's another thing to actually detect it. There's even a comment that says "Don't make this a standard until there's a reference implementation" and I get why they say that.

**Comment by MikeFair** [comment]
@Miosss

The other thing about why it matters to expose this at the API level is because we want the part where an application creates a JSON string of their data to get skipped.

What creates the efficiency is when my application can take an array of floats, that it has already created, and call the API to "make this part of the JSON I'm encoding" directly; avoiding myFloatArray.toJSONString().toUBJ()

Conversely, the exact opposite has to happen on the other side.
It makes a big difference if the application can somehow skip the part where it uses a JSON String to marshall the data back and forth.
In otger words, skip the part it does myDecodedData = ubjSourceString.toJSONString().toMyLocalAppType().

**Comment by MikeFair** [comment]
So to complete the thought; the comment was about addressing the needs for strongly typed languages to use the strongly typed efficient containers that exist natively in conjuction with JSON, UBJ, and the end user API.  Though I was really just focused on the encoder/decoder my comment. The broader goal was "How can applications take advantage of these optimized containers while skipping the to/fromJsonString() exercise".

**Comment by Miosss** [comment]
> In otger words, skip the part it does myDecodedData = ubjSourceString.toJSONString().toMyLocalAppType().

Thats obvious. It is you, who mentioned JSONArray.Add etc.

In API I imagine ubjValue.Add(array_of_int) end of story. No segments, just array (can be of mixed types).

**Comment by MikeFair** [comment]
The point of the psuedocode and segment discussion is to make/justify a
resolution/movement that:
1) JSON "Containers", by definition, are mixed type. Right? So if we use
the word Array, Object, or Container; then we must, by definition be
including mixed type things.
2) Therefore an "Optimized Container" solution, requires a mixed type
solution.  Proposing a container optimization that must degrade to a fully
expanded direct JSON translation (as some proposals do) cannot be
acceptable as the general solution.  We can call it a special use case
Container.  But it's not a generic "Optimized Container" solution. (I don't
think I'm saying anything new, just clarifying some terms.)
3) What's clear is that one way or another Strongly Typed data has to be
dealt with in blocks/regions where their sizes are known (like little
voxels, if you know that reference).  So by extension that means "Optimized
Containers" (Arrays and Objects) will have to make use of Strongly Typed
regions of data.
4)  It'd be good to have a name for these strongly typed regions of data.
They can't be called "Container" because those (Array and Object) are mixed
type. I called them "Segment"; would "Block", "Region", or "Section" be a
better term?  These blocks are not containers themselves, they are
something containers use/reference.

** This doesn't mean we can't have an optimization specific to containers
that only consist of a single type; just that they can't be the whole
optimized container solution.  However I think everyone already agrees with
this and that a good mixed type solution will handle the single type
gracefully.

I'm clarifying this to point out that these repeatable headers/optimization
descriptors are actually creating a new kind of UBJ specific thing that a
mixed type container can be composed from; this is separate from actually
_being_ a container on its own.  This means an API will need to give a
client a way to deal with them.

>> In other words, skip the part where it does myDecodedData =
ubjSourceString.toJSONString().toMyLocalAppType().
>
> Thats obvious. It is you, who mentioned JSONArray.Add etc.
>
> In API I imagine ubjValue.Add(array_of_int) end of story. No segments,
just array (can be of mixed types).

I think you meant to refer to the other psuedocode example?  The psuedocode
you provided in the reply seems to be about adding data to be encoded to
UBJ; as opposed to the quote which is about getting data from UBJ data into
your app.  And, yes, in this case I am thinking that someone is using a UBJ
library that interacts as directly as possible with the data structures in
their application instead of hand encoding/decoding the UBJ in their own
code.

When I brought up jsonArray.Add() it was to show how strongly typed data
can be successfully merged into a mixed type JSON Array (as a strongly
typed thing - or that the Array can use these things internally), remaining
performant, and easy to use.

Certainly it's not the only way, or even the definitive way; just _a way_
to show that it can be done to clear the path for a mixed type optimized
container solution.

We sidestep the issue by inventing this new thing that is neither a
container, nor a single value; it's a strongly typed set of values (and it
requires a name).
